packages/venia-concept/src/components/SearchBar/autocomplete.js

```diff
@@ -1,154 +1,94 @@
-import React, { Component } from 'react';
-import PropTypes from 'prop-types';
+import React, { useCallback, useEffect } from 'react';
+import { bool, shape, string } from 'prop-types';
 import debounce from 'lodash.debounce';
-import { Query } from 'src/drivers';
-
-import classify from 'src/classify';
-import { loadingIndicator } from 'src/components/LoadingIndicator';
-import SuggestedCategories from './suggestedCategories';
-import SuggestedProducts from './suggestedProducts';
-import PRODUCT_SEARCH from '../../queries/productSearch.graphql';
+import { useFieldState } from 'informed';
+import { useApolloContext, useQueryResult } from '@magento/peregrine';
 
+import { mergeClasses } from 'src/classify';
+import PRODUCT_SEARCH from 'src/queries/productSearch.graphql';
+import Suggestions from './suggestions';
 import defaultClasses from './autocomplete.css';
 
 const debounceTimeout = 200;
-const suggestedCategoriesLimit = 4;
-const suggestedProductsLimit = 3;
-
-class SearchAutocomplete extends Component {
-    static propTypes = {
-        classes: PropTypes.shape({
-            root: PropTypes.string,
-            statusContent: PropTypes.string
-        }),
-        searchQuery: PropTypes.string.isRequired,
-        autocompleteVisible: PropTypes.bool,
-        executeSearch: PropTypes.func.isRequired,
-        updateAutocompleteVisible: PropTypes.func.isRequired
-    };
 
-    constructor(props) {
-        super(props);
-        this.state = {
-            isQueryUpdating: false,
-            autocompleteQuery: props.searchQuery || ''
-        };
+const Autocomplete = props => {
+    const { visible } = props;
+    const { data, dispatch, error, loading } = useQueryResult();
+    const client = useApolloContext();
+    const { value } = useFieldState('search_query');
+    const classes = mergeClasses(defaultClasses, props.classes);
+    const rootClassName = visible ? classes.root_visible : classes.root_hidden;
+    const valid = value && value.length > 2;
+    let message = '';
+
+    if (error) {
```

Comment:
This is an interesting stylistic discussion. @zetlen and I have had a few conversations about it as well.

With a class component, the various responsibilities of `render` can be extracted to getters or methods (including lifecycle methods). With function components, everything that can't be extracted into a utility or custom hook happens inside the one function body, which can get messy when it starts to get complicated.

Ideally, a pattern would start to emerge for the order of operations inside the function body—a convention, not a rule. Here, I've tried to come up with one:

1. variables & destructuring (setup)
2. callback definitions (`useCallback`)
3. side effects (`useEffect`)
4. output (`return`)

This order of operations is something that's intuitive to **me**, but I don't know if it's necessarily intuitive to others. I also think it's aesthetically clean because it tends to collate one-liners at the top, with no extra whitespace between lines.

Let me try group them more by purpose, though, as you've suggested here, and see how it feels to me.

Reply:
> can get messy when it starts to get complicated.

True, in real apps this stuff will be all intertwined.
I like the convention you outlined but I'd be interested to see how it felt when you switched it up.

Reply:
This is a place where we should let a hundred flowers bloom. I understand both the desire to colocate pieces of related data and logic, and the desire to have a "view model construction" section and then a JSX section of a render method.

I think both are okay! It's hard to know how functionality will end up composing, and hooks make it possible for us to recombine things experimentally inside a block, instead of weaving throughout a class.

So I think this is good for now, but I endorse @supernova-at's message as well.

Reply:
@zetlen Agreed. Very soon, we're going to have multiple authors working on hooks and hook components, and I expect this is where we'll see the greatest amount of diversity in terms of style. I hope that a pattern emerges, but I don't think it has to be rigid.

That said, I've already found that I'm adjusting my own style to group things closer to the way @supernova-at is thinking. 🙂